keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+Example:
+
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] controller “msg”=”Updating ready status” "core/node": {"name": "node-1"} "core/pod": {"name": "nginx-1", "namespace": "default"} "controller": {"status": false}
+```
+
+### Logging configuration
+
+ 
+A new flag will be introduced to all kubernetes components allowing users to select their preferred logging format and maintain backward compatibility.
+Proposed flag `--logging-format`, with 3 values:
+* `legacy` for old logging format
+* `json` for json production format
+* `dev` for development format
+
+Setting this flag will select particular logger implementation from registry. Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
+
+We acknowledge that extension via code is not the best, but should be sufficient for now. 
+
+All logs from the `json` and `dev` loggers should go to standard output.  Out of the existing klog command line arguments `json` and `dev` logger should support ‘-v’ flag.  
+
+### Kubernetes log schema
+
+Introducing structured logging API is not enough to address one of the biggest issues with log creation or analysis, that is inconsistent metadata. Size of Kubernetes codebase could easily result in drifting naming conventions. As part of this proposal we want to propose a mechanism for ensuring consistency of metadata. To achieve this we will introduce a way to define a schema for subset of commonly used fields in kubernetes. This schema should improve reusability of fields and provide discovery mechanism for logging backend vendors.
+
+Introducing this schema will include three things: 
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
+* Validation mechanism ensuring schema is not bypassed by lower level logging API
+
+TODO(serathius): Propose how definition format should look like, probably yaml similar to https://github.com/elastic/ecs/tree/master/schemas
+
+### Migration
+ 
+Migration to new structured logging API will be divided into 4 stages:
+
+#### Stage 1 / Alpha
+
+This stage needs to be performed in a single release of Kubernetes. When finished for every **existing** log entry produced in legacy output format there should be a corresponding log entry in json/dev format containing similar information possibly with some additional context information attached to it.
+When using the legacy log format after this stage Kubernetes control plane components should produce exactly the same logs as before this stage. 
```

Comment:
I think this needs code examples to demonstrate. I can imagine what this probably looks like, but I'd like this to be as specific as possible as this is very crucial to the entire effort.